janusgraph-core/src/main/java/org/janusgraph/core/attribute/GeoshapeHelper.java

```diff
@@ -107,9 +153,47 @@ public int size(Shape shape) {
                     return new Geoshape.Point(shape.getBoundingBox().getMinY(), shape.getBoundingBox().getMinX());
                 else
                     return new Geoshape.Point(shape.getBoundingBox().getMaxY(), shape.getBoundingBox().getMaxX());
+            case LINE:
+                p = ((BufferedLineString) shape).getPoints().get(position);
+                break;
+            case MULTIPOINT:
+                p = ((ShapeCollection<Point>) shape).getShapes().get(position);
+                break;
+            case MULTILINESTRING:
+                cpt = 0;
+                final Iterator<BufferedLineString> itLines = ((ShapeCollection<BufferedLineString>) shape).getShapes().iterator();
+                while (itLines.hasNext()) {
+                    final Iterator<Point> itPoints = itLines.next().getPoints().iterator();
+                    while (itPoints.hasNext()) {
+                        if (cpt == position) {
+                            p = itPoints.next();
+                            break;
+                        }
+                        cpt++;
+                        itPoints.next();
+                    }
+                    if (p != null) {
+                        break;
+                    }
+                }
+                break;
+            case GEOMETRYCOLLECTION:
+                final Iterator<Shape> itShapes = ((ShapeCollection<Shape>) shape).getShapes().iterator();
+                cpt = position;
+                while (itShapes.hasNext()) {
+                    final Shape internShape = itShapes.next();
+                    final int nbElement = size(internShape);
+                    if (cpt < nbElement) {
+                        return getPoint(new Geoshape(internShape), cpt);
+                    } else {
+                        cpt -= nbElement;
+                    }
+                }
+                break;
             default:
                 throw new IllegalStateException("getPoint(int) not supported for type: " + getType(shape));
         }
```

Comment:
`if (p == null) throw new ArrayIndexOutOfBoundsException("Invalid position: " + position);` for completeness.